README.md
# PoseCorrecter

Check if you got an ergonomic working pose behind your laptop. This will be decided by the position of your nose in front of the webcam. You could use the Posecorrector to send notfications when you aren't sitting in the correct pose.
<p>&nbsp;</p>

## Why is it useful
More and more people have all sorts of pain because of the pose they take-on when working behind their computer. This api would be able to send notifications when you unconciously (or conciously) would take on a bad sitting position.
<p>&nbsp;</p>

## Getting Started

Follow these instructions to get your own version running.

1. Clone or download this repository.
2. Navigate to the api folder `cd api` Install Node Package Manager `npm install`
3. Navigate to the root folder `cd ../` in the terminal and start the docker-compose. Install docker [here](https://docs.docker.com/desktop/) if you haven't already. 
```
Docker-compose build
```
```
Docker-compose up
```

### Posenet
Posenet uses Machine learning to predict where 17 parts of your body would be. For more information refer to [here](https://ml5js.org/reference/api-PoseNet/). <br><br>
Start up the index.html file (for example on port 5500). Accept the use of camera if your browser asks permission. In the deploy.js file you can see the xWaarde and yWaarde for the nose in the 'classifypose' function. Try to export these values because i did not succeed in this, i will be trying again later. For now dummydata in 'data.js' is used for 2 positsions of the nose.
<br>


### Endpoints:

Access the endpoints like this:

- `GET /sessions` Get all the sessions, these are the feedback on your pose
- `GET /measurements` Get all the measurements, the positions of your nose with the id of the session

- `GET /sessions/:uuid` Get a specific session
- `GET /measurements/:uuid` Get a specific measurement

- `POST /sessions` Post a session or sessions (uuid, feedback)
- `POST /measurements` Post a measurement or measurements (uuid, xWaarde, yWaarde, session_id)

- `PATCH /sessions/:uuid` Patch/update a specific session by uuid
- `PATCH /measurements/:uuid` Patch/update a specific measurement by uuid

- `DELETE /sessions/:uuid` Delete a specific session by uuid
- `DELETE /measurements/:uuid` Delete a specific measurement by uuid

### Sessions:
The sessions contain the information if a position is good or bad (feedback). They would be linked to the measurements.
### Measurements:
The measurements contain the information about the x and y position of the nose in front of the screen. They also contain a session_id to know if the position is good or bad.




<p>&nbsp;</p>


## Running the tests

Navigate back to the api folder `cd api` and run the test `npm test`. <br> There are 3 tests written. (helpers, end-to-end, integration) More information about the tests is visible by the test itself.
<p>&nbsp;</p>

## Where can i find help

Use the issue and discussion tab if you would like to receive help. More information can be found in the [Contributing guidelines](https://github.com/Toon3690/PoseCorrecter/blob/main/CONTRIBUTING.md).

<p>&nbsp;</p>

## Project status
The project is still under construction.

<p>&nbsp;</p>

## Github flow
We make use of the Github Flow. This is because it is a small project, currently being worked on by one person. Also because we can deploy it so much faster. Besides it is much faster to fix bugs or even implement features. All the changes happen through pull requests so others can review it.
If you want to work on the project, get familiar with the [Github flow](https://guides.github.com/introduction/flow/).

<p>&nbsp;</p>


## Built With

* [JEST](https://jestjs.io/en/)
* [Docker](https://docs.docker.com/)
* [Knex](http://knexjs.org/)
* [NPM](https://www.npmjs.com/)
* [Tabelplus](https://tableplus.com/)
* [Fork](https://git-fork.com/)
## Contributing

Please read [CONTRIBUTING.md]() for details on our code of conduct, and the process for submitting pull requests to us.

<p>&nbsp;</p>

## Changelog

In the Changelog all the made changes are stored.

<p>&nbsp;</p>

## Authors

*  [**Toon Raskin**](https://github.com/Toon3690)
*  Help from Jav Everaert

<p>&nbsp;</p>

## License

This project is licensed under the MIT License - see the [LICENSE.md](https://github.com/Toon3690/PoseCorrecter/blob/main/LICENSE) file for details




CHANGELOG.md
# Changelog
Changes are documented here, with the last made changes on top.


## V 1.2.0
### Extra
```
-  Deploy.js
-  index.html
```


## V 1.1.0
### Extra
```
-  CODE-OF-CONDUCT.md
-  CONTRIBUTING.md
-  CHANGELOG.md
```
## V 1.0.0

### Endpoints
```
- `PATCH /sessions/:uuid` Patch/update a specific session by uuid
- `PATCH /measurements/:uuid` Patch/update a specific measurement by uuid

- `DELETE /sessions/:uuid` Delete a specific session by uuid
- `DELETE /measurements/:uuid` Delete a specific measurement by uuid
```
### Tests
```
-  End-To-End test
```
### Extra
```
-  Deploy.js and index.html for position of nose
```
## V 0.6.0

### Endpoints
```
- `GET /sessions/:uuid` Get a specific session
- `GET /measurements/:uuid` Get a specific measurement

- `POST /sessions` Post a session or sessions (uuid, feedback)
- `POST /measurements` Post a measurement or measurements (uuid, xWaarde, yWaarde, session_id)
```
### Tests
```
-  Integration test updated
```
### Extra
```
-  DatabaseHelper.js
```


## V 0.1.0

### Endpoints
```
- `GET /sessions` Get all the sessions, these are the feedback on your pose
- `GET /measurements` Get all the measurements, the positions of your nose with the id of the session
```
### Tests
```
-  Integration test
```
### Extra
```
-  Helpers.js
-  Dockerfiles
```



api/src/__tests__/helpers.test.js
const Helpers = require('../utils/helpers');
const uuid = Helpers.generateUUID();

describe('uuid validator test', () => {
    test('test if it is a string and test if string is not empty', () => {
        console.log(uuid);
        expect(Helpers.validateUUID("ezfzef3151ef")).toBeFalsy();
        expect(Helpers.validateUUID()).toBeFalsy();
    });
    test('test if an empty string', () => {
        expect(Helpers.validateUUID("")).toBeFalsy();
    });
});

 describe('poseTest', () => {
    test('test if values return correct poseCode', () => {
        expect(Helpers.checkPosture(200, 33)).toEqual(1);
        expect(Helpers.checkPosture(320, 240)).toEqual(4);
        expect(Helpers.checkPosture(151, 240)).toEqual(2);
        expect(Helpers.checkPosture(350, 2)).toEqual(3);
        expect(Helpers.checkPosture(-1, 240)).toEqual(5);
    })
}); 

api/src/__tests__/integration.test.js
const supertest = require('supertest');
const app = require('../server');
const Helpers = require('../utils/helpers');

const request = supertest(app);

const UUID = Helpers.generateUUID();

//Test sessions
describe('GET /measurements endpoint', () => {

    test('check if response is 200 (OK)', async (done) => {
        try {
            await request.get('/measurements')
                .expect(200)
                .then((res) => {
                    done();
                    console.log("got measurements");
                });
        } catch (e) {
            if (e) console.log(e);
            done();
        }
    });

    test('check if response is undefined', async (done) => {

        try {
            await request.get(`/measurements/${UUID}`)
                .expect(200)
                .then((res) => {
                    expect(res.body.res[0]).not.toBeDefined;
                    done();
                    console.log("undefined");
                });
        } catch (e) {
            if (e) console.log(e); 
            done();
        }
    });
});

//Test measurements
describe('POST /measurements endpoint', () => {
    test('check if response is 201 (Created)', async (done) => {
        try {
            await request.post('/measurements')
                .send({
                    "uuid": UUID,
                    xWaarde: 200,
                    yWaarde: 109,
                    session_id: 5
                })
                .expect(201)
                .then((res) => {
                    done();
                    console.log("201 code POST");
                });
        } catch (e) {
            if (e) console.log(e);
            done(e)
            done();
        }
    })

    test('check if response is 201 (Created), when no uuid is send', async (done) => {
        try {
            await request.post('/measurements')
            .send({
                xWaarde: 800,
                yWaarde: 700,
                session_id: 5
            })
                .expect(201)
                .then((res) => {
                    done();
                    console.log("201 code POST");
                });
        } catch (e) {
            if (e) console.log(e);
            done(e)
            done();
        }
    })
});

api/src/__tests__/end-to-endTest.test.js
//for end-to-end testing 

// voeg een session toe via de endpoint
// check in de database of het is toegevoegd
// probeer via een get request terug op te halen
// update
// check of update is doorgevoerd

const supertestETE = require('supertest');
const httpETE = require('http');

const db = require('../server');

const values = require("../data");

const appETE = require('../server');
const requestETE = supertestETE(appETE);
const HelpersETE = require('../utils/helpers');

let ETE_uuid;



describe('test end-to-end sessions', () => {

    test('Add a session and check if POST succeeds', async (done) =>{
        const response = await requestETE.post('/sessions').send({feedback: "wonderful"});
        expect(response.status).toBe(201)
        expect(response.body).toHaveProperty("uuid", response.body.uuid)
        ETE_uuid = response.body.uuid
        done();
    })

    test('Check if the GET request finds this session', async (done) =>{
        const response = await requestETE.get(`/sessions/${ETE_uuid}`).send()
        expect(response.status).toBe(200)
        expect(response.body.res[0]).toHaveProperty("uuid")
        done();
    })

    test('Update the session and check if PATCH succeeds', async (done) =>{
        const response = await requestETE.patch(`/sessions/${ETE_uuid}`).send({ feedback: "disaster" })
        expect(response.status).toBe(201)
        expect(response.body.res[0]).toHaveProperty("uuid")
        expect(response.body.res[0]).toHaveProperty("feedback", "disaster")
        done();
    })

    test('Check if GET request detects change', async (done) =>{
        const response = await requestETE.get(`/sessions/${ETE_uuid}`).send()
        expect(response.status).toBe(200)
        expect(response.body.res[0]).toHaveProperty("uuid")
        expect(response.body.res[0]).toHaveProperty("feedback", "disaster")
        done();
    })  

      test('DELETE the session and check if succeeds', async (done) =>{
        const response = await requestETE.delete(`/sessions/${ETE_uuid}`).send()
        expect(response.status).toBe(205)
        done();
    })

    test('if GET request is empty when getting the deleted session', async (done) =>{
        const response = await requestETE.get(`/sessions/${ETE_uuid}`).send()
        expect(response.status).toBe(200);
        expect(response.body.res[0]).toBeUndefined();  
        console.log("hier ook geraakt");
        done();
    })   
});



describe('test end-to-end measurements', () => {

    test('Add a measurement and check if POST succeeds', async (done) =>{
        const response = await requestETE.post('/measurements').send({xWaarde: values[0][1].x, yWaarde: values[0][1].y});
        expect(response.status).toBe(201)
        expect(response.body).toHaveProperty("uuid")
        ETE_uuid = response.body.uuid
        done();
    })
    
    test('Check if the GET request finds this measurement', async (done) =>{
        const response = await requestETE.get(`/measurements/${ETE_uuid}`).send()
        expect(response.status).toBe(200)
        expect(response.body.res[0]).toHaveProperty("uuid")
        done();
    })

    test('Update the measurement and check if PATCH succeeds', async (done) =>{
        const response = await requestETE.patch(`/measurements/${ETE_uuid}`).send({xWaarde: values[0][2].x, yWaarde: values[0][2].y, session_id: HelpersETE.checkPosture(values[0][2].x, values[0][2].y)})
        expect(response.status).toBe(201)
        expect(response.body.res[0]).toHaveProperty("uuid")
        expect(response.body.res[0]).toHaveProperty("xWaarde", values[0][2].x)
        expect(response.body.res[0]).toHaveProperty("yWaarde", values[0][2].y)
        expect(response.body.res[0]).toHaveProperty("session_id", "1")
        done();
    })

    test('if PATCH request fails when send no uuid', async (done) =>{
        const response = await requestETE.patch(`/measurements`).send({xWaarde: values[0][2].x, yWaarde: values[0][2].y, session_id: HelpersETE.checkPosture(values[0][2].x, values[0][2].y)})
        expect(response.status).toBe(404)
        done();
    })

     test('Check if GET request detects change', async (done) =>{
        const response = await requestETE.get(`/measurements/${ETE_uuid}`).send()
        expect(response.status).toBe(200)
        expect(response.body.res[0]).toHaveProperty("uuid")
        expect(response.body.res[0]).toHaveProperty("xWaarde", values[0][2].x)
        expect(response.body.res[0]).toHaveProperty("yWaarde", values[0][2].y)
        expect(response.body.res[0]).toHaveProperty("session_id", "1")
        done();
    })   

     test('DELETE the measurement and check if succeeds', async (done) =>{
        const response = await requestETE.delete(`/measurements/${ETE_uuid}`).send()
        expect(response.status).toBe(205)
        done();
    })

    test('if GET request is empty when getting the deleted measurement', async (done) =>{
        const response = await requestETE.get(`/measurements/${ETE_uuid}`).send()
        expect(response.status).toBe(200);
        expect(response.body.res[0]).toBeUndefined();  
        done();
    })  
});



api/src/utils/helpers.js
const {
  v1: uuidv1
} = require('uuid');

const {
  v4: uuidv4
} = require('uuid');

const Helpers = {

  // Generate a uuid timebased
  generateUUID: () => {
    const uuid = uuidv1();
    return uuid;
  },

  // Generate a uuid at random
  generateUUID2: () => {
    const uuid = uuidv4();
    return uuid;
  },

  // Check if the uuid is actually a valid uuid
  validateUUID: (uuid) => {
    if (uuid == undefined) {
      return false
    } else {
      if (uuid.length > 0 && typeof uuid == uuid) {
        return true
      } else {
        return false
      }
    }
  },

  // returns number based on the position of x and y (should be edited, maybe working with a circle)
  checkPosture: (xWaarde, yWaarde) => {
    if (xWaarde < 320 && xWaarde >= 0 && yWaarde < 240 && yWaarde >= 0) {
      return 1
    } else if (xWaarde < 320 && xWaarde >= 0 && yWaarde >= 240 && yWaarde <= 480) {
      return 2
    } else if (xWaarde >= 320 && xWaarde <= 640 && yWaarde < 240 && yWaarde >= 0) {
      return 3
    } else if (xWaarde >= 320 && xWaarde <= 640 && yWaarde >= 240 && yWaarde <= 480) {
      return 4
    } else {
      return 5
    }
  },
};

module.exports = Helpers;

api/src/utils/DatabaseHelper.js
const Helpers = require('./helpers.js');

// Connection with database via knex
const pg = require('knex')({
    client: 'pg',
    version: '9.6',
    searchPath: ['knex', 'public'],
    connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
});

const DatabaseHelper = {
    async initialiseTables() {

        // Create measurements table
        await pg.schema.hasTable('measurements').then(async (exists) => {
            if (!exists) {
                await pg.schema
                    .createTable('measurements', (table) => {
                        table.increments();
                        table.uuid('uuid');
                        table.json('xWaarde');
                        table.json('yWaarde');
                        table.string('session_id');

                        table.timestamps(true, true);
                    })
                    .then(async () => {
                        console.log('created measurements');
                    });
            } else {
                console.log("measurements already created");
            }
        });

        // Create table sessions and add the 5 basic sessions
        await pg.schema.hasTable('sessions').then(async (exists) => {
            if (!exists) {
                await pg.schema
                    .createTable('sessions', (table) => {
                        table.increments();
                        table.uuid('uuid');
                        table.string('feedback');
                        table.timestamps(true, true);
                    })
                    .then(async () => {
                        console.log('created sessions');
                        const uuid = Helpers.generateUUID();

                        await pg.table('sessions').insert([
                            {uuid, feedback: `Good` },
                            {uuid, feedback: `Bad` },
                            {uuid, feedback: `Worst` },
                            {uuid, feedback: `Bend` },
                            {uuid, feedback: `Not registered` }
                        ]);
                    });

            } else {
                console.log("sessions already created");
            }
        });
    }
}

//test
module.exports = DatabaseHelper, pg;

api/src/server.js
const express = require("express");
const bodyParser = require("body-parser");
const http = require("http");

const Helpers = require('./utils/helpers.js');
const DatabaseHelper = require('./utils/DatabaseHelper');

const app = express();
http.Server(app);

app.use(bodyParser.json());
app.use(
    bodyParser.urlencoded({
        // to support URL-encoded bodies
        extended: true
    })
);


// Initialise the tables via databasehelper function
DatabaseHelper.initialiseTables();

// Connection with database via knex
const pg = require('knex')({
    client: 'pg',
    version: '9.6',
    searchPath: ['knex', 'public'],
    connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
});



/**
 * Get all the sessions and the data from it
 * @params: none
 * @returns: json or statuscode 400 (Bad request)
 */
app.get('/sessions', async (req, res) => {
    
    const result = await pg
        .select('*')
        .from('sessions')
    if (result) {
        res.json({
            res: result
        })
    } else {
        res.send(400)
    }
});

/**
 * Get all the measurements and the data from it
 * @params: none
 * @returns: json or statuscode 400 (Bad request)
 */
app.get('/measurements', async (req, res) => {

    const result = await pg
        .select('*')
        .from('measurements')
    if (result) {
        res.json({
            res: result
        })
    } else {
        res.send(400)
    }
});



/**
 * Get one session and the data from it
 * @params: uuid - of the required session
 * @returns: json and statuscode 200 (OK) or statuscode 400 (Bad Request) or 404 (Not Found)
 */
app.get('/sessions/:uuid', async (req, res) => {

    // check if uuid is valid
    if (Helpers.validateUUID(!req.params.uuid)) {
        res.status(400);
    } else {
        const result = await pg
            .select('*')
            .from('sessions')
            .where({
                uuid: req.params.uuid
            })
        .then((result) => {
            res.status(200);
            res.json({
                res: result
            });
        })
        .catch((e) => {
            console.log(e);
            res.status(404);
        });
    }
});

/**
 * Get one measurement and the data from it
 * @params: uuid - of the required session
 * @returns: json and statuscode 200 (OK) or statuscode 400 (Bad Request) or 404 (Not Found)
 */
app.get('/measurements/:uuid', async (req, res) => {

    // check if uuid is valid
    if (Helpers.validateUUID(!req.params.uuid)) {
        res.status(400);
    } else {
        const result = await pg
            .select('*')
            .from('measurements')
            .where({
                uuid: req.params.uuid
            })
            .then((result) => {
                res.status(200);
                res.json({
                    res: result
                });
            })
            .catch((e) => {
                console.log(e);
                res.status(404);
            });
    }
});




/**
 * POST or create a session
 * @params: uuid, feedback
 * @returns: json with uuid or statuscode 201 (Created) or 400 (Bad Request)
 */
app.post('/sessions', async (req, res) => {

    // Make a uuid if not specified
     let uuid;
    if(!req.body.uuid){
         uuid = Helpers.generateUUID();
    }else{
         uuid = req.body.uuid
    } 

    const result = await pg
        .insert({
            uuid: uuid,
            feedback: req.body.feedback,
        })
        .into("sessions")
        .then(() => {
            res.status(201);
            res.json({
                uuid: uuid
            });
        })
        .catch((e) => {
            console.log(e);
            res.status(400);
        });
});

/**
 * POST or create measurements
 * @params: uuid, xWaarde, yWaarde, session_id
 * @returns: json with uuid or statuscode 201 (Created) or 400 (Bad Request)
 */
app.post('/measurements', async (req, res) => {

    // Make a uuid if not specified
    let uuid;
    if(!req.body.uuid){
         uuid = Helpers.generateUUID();
    }else{
         uuid = req.body.uuid
    }

        const result = await pg
        .insert({
            uuid: uuid,
            xWaarde: req.body.xWaarde,
            yWaarde: req.body.yWaarde,
            session_id: Helpers.checkPosture(req.body.xWaarde, req.body.yWaarde)
        })
        .table("measurements")
        .then(() => {
            res.status(201);
            res.json({
                uuid: uuid,
                xWaarde: req.body.xWaarde
            });
            res.send();
        })
        .catch((e) => {
            console.log(e);
            res.status(400);
        });
});





/**
 * PATCH or update session
 * @params: uuid 
 * @returns: json and statuscode 201 (Created) or 400 (Bad Request) or 404(Not Found)
 */
app.patch('/sessions/:uuid', async (req, res, done) => {

    // Check if uuid is valid
    if (Helpers.validateUUID(!req.params.uuid)) {
        res.status(400);
    } else {
        const result = await pg
            .update(req.body)
            .from('sessions')
            .where({
                uuid: req.params.uuid
            })
            .returning('*')
        if (result) {
            res.status(201)
            res.json({
                res: result
            })
        } else {
            res.status(400)
        }
    }
})


/**
 * PATCH or update measurement
 * @params: uuid 
 * @returns: json and statuscode 201 (Created) or 400 (Bad Request) or 404(Not Found)
 */
app.patch('/measurements/:uuid', async (req, res, done) => {

    // Check if uuid is valid
    if (Helpers.validateUUID(!req.params.uuid)) {
        res.status(400);
    } else {
        const result = await pg
            .update(req.body)
            .from('measurements')
            .where({
                uuid: req.params.uuid
            })
            .returning('*')
        if (result) {
            res.status(201)
            res.json({
                res: result
            })     
        } else {
            res.status(404)
        }
    }
})




/**
 * DELETE a session
 * @params: uuid
 * @returns: json and statuscode 205 (Reset content) or 400 (Bad Request) or 404(Not Found)
 */
app.delete('/sessions/:uuid', async (req, res) => {

    // Check if uuid is valid
    if (Helpers.validateUUID(!req.params.uuid)) {
        res.status(400);
    } else {
        const result = await pg
            .from('sessions')
            .where({
                uuid: req.params.uuid
            })
            .del('*')
        if (result) {
            res.status(205)
            res.json({
                res: result
            })
            
        } else {
            res.status(404)
        }
    }
});


/**
 * DELETE a session
 * @params: uuid
 * @returns: json and statuscode 205 (Reset content) or 400 (Bad Request) or 404(Not Found)
 */
app.delete('/measurements/:uuid', async (req, res) => {

    // Check if uuid is valid
    if (Helpers.validateUUID(!req.params.uuid)) {
        res.status(400);
    } else {
        const result = await pg
            .from('measurements')
            .where({
                uuid: req.params.uuid
            })
            .del('*')
        if (result) {
            res.status(205)
            res.json({
                res: result
            })
            
        } else {
            res.status(404)
        }
    }
});


module.exports = app, pg;